module/x/gravity/keeper/signer_set.go

```diff
@@ -0,0 +1,51 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/peggyjv/gravity-bridge/module/v3/x/gravity/types"
+)
+
+// TODO here we should check the contents of the validator set against
+// the store, if they differ we should take some action to indicate to the
+// user that bridge highjacking has occurred
+func (k Keeper) signerSetExecuted(ctx sdk.Context, nonce uint64) {
+	otx := k.GetOutgoingTx(ctx, types.MakeSignerSetTxKey(nonce))
+	if otx == nil {
+		k.Logger(ctx).Error("Failed to clean signer sets",
+			"signer set nonce", nonce)
+		return
+	}
+
+	completedSignerSetTx, _ := otx.(*types.SignerSetTx)
+	k.setLastObservedSignerSetTx(ctx, *completedSignerSetTx)
+	k.CompleteOutgoingTx(ctx, completedSignerSetTx)
```

Comment:
OK, so I believe we have a bug here. `CompleteOutgoingTx` deletes the OTX and stores a completed TX. In `BeginBlocker` we run `pruneSignerSetTxs` which will remove all the signer set OTXs *except* the last one. This is because `createSignerSetTx` checks to see if an OTX for a signer set exists by calling `GetLatestSignerSetTx`, and if it doesn't find one, it creates one and exits. The net effect of this would be creating a signer set tx every single time after one completes. We need to either account for leaving one around, or adjust the creation of signer set TXs to handle looking at completed ones.

Reply:
Looking back at this I don't think it's a bug?

> we run `pruneSignerSetTxs` which will remove all the signer set OTXs except the last one.

It will remove all signer sets before the last *observed* one. Keep in mind `LastSignerSetTx` and `LastObservedSignerSetTx` are distinct. 

So the order would be
1. Signer set is updated in Gravity contract, and observed by the chain
2. The event handler runs `signerSetExecuted` which deletes the OTX, saves it as completed, and *also saves it as the last observed signer set tx*
3. BeginBlocker runs `createSignerSetTx` which will see the last tx was deleted and create a new one (this is weird because it bumps the nonce for basically no reason, so still worth letting pruning in abci handle it?).
4. BeginBlocker runs `pruneSignerSetTxs` which only deletes signer set txs with nonces lower than the last *observed* signer set tx.

Basically the only effect calling `CompleteOutgoingTx` inside of `signerSetTxExecuted` is that a new signer set tx is immediately created that is unlikely to be called for by a power differential, so it's still cleaner to let pruning eventually delete the associated `OutgoingTx` imo. With the current change, it should only set it as completed, and then eventually when another tx has been observed it will be pruned after the window passes.

I considered whether to have pruning delete the last observed's associated OTX instead of only nonces lower than it, but this would throw off the guarantee of always having a path of signer sets to run one after the other if getting from set A to set Z can't be done in one transaction.

Also worth noting that the only thing that looks at the last observed signer set tx is the pruning method; the relayer doesn't use it.